File: tamiltts/mlx/normalize.py
"""Text normalization (verbalization) front-end — run BEFORE char tokenization.

Converts written forms the char-level model never saw into pronounceable, in-vocab Tamil:
  - acronyms / Latin-letter runs -> spelled out as Tamil letter-names, with periods removed
    (periods otherwise read as sentence-final pauses, making "I.I.T." sound choppy)
  - common symbols -> Tamil words
  - digit runs -> spoken digit-by-digit in Tamil (full cardinal number-words = future work)

Mirrored 1:1 by the Rust SDK (rust/src/normalize.rs); keep the two in sync.
"""
from __future__ import annotations

import re

# English letter names rendered in Tamil script (for spelling acronyms like IIT, MLX, TTS).
LETTER = {
    "a": "ஏ", "b": "பி", "c": "சி", "d": "டி", "e": "ஈ", "f": "எஃப்", "g": "ஜி",
    "h": "ஹெச்", "i": "ஐ", "j": "ஜே", "k": "கே", "l": "எல்", "m": "எம்", "n": "என்",
    "o": "ஓ", "p": "பி", "q": "கியூ", "r": "ஆர்", "s": "எஸ்", "t": "டி", "u": "யூ",
    "v": "வி", "w": "டபிள்யூ", "x": "எக்ஸ்", "y": "வை", "z": "செட்",
}

DIGIT = ["பூஜ்ஜியம்", "ஒன்று", "இரண்டு", "மூன்று", "நான்கு", "ஐந்து", "ஆறு", "ஏழு", "எட்டு", "ஒன்பது"]

# --- Tamil cardinal number words (Indian place-value grouping) ---
_TEENS = ["பத்து", "பதினொன்று", "பன்னிரண்டு", "பதிமூன்று", "பதினான்கு", "பதினைந்து",
          "பதினாறு", "பதினேழு", "பதினெட்டு", "பத்தொன்பது"]
_TENS_ST = ["", "பத்து", "இருபது", "முப்பது", "நாற்பது", "ஐம்பது", "அறுபது", "எழுபது", "எண்பது", "தொண்ணூறு"]
_TENS_CO = ["", "", "இருபத்தி", "முப்பத்தி", "நாற்பத்தி", "ஐம்பத்தி", "அறுபத்தி", "எழுபத்தி", "எண்பத்தி", "தொண்ணூற்றி"]
_HUND_ST = ["", "நூறு", "இருநூறு", "முந்நூறு", "நானூறு", "ஐந்நூறு", "அறுநூறு", "எழுநூறு", "எண்ணூறு", "தொள்ளாயிரம்"]
_HUND_CO = ["", "நூற்றி", "இருநூற்று", "முந்நூற்று", "நானூற்று", "ஐந்நூற்று", "அறுநூற்று", "எழுநூற்று", "எண்ணூற்று", "தொள்ளாயிரத்து"]
_THOU_ST = ["", "ஆயிரம்", "இரண்டாயிரம்", "மூவாயிரம்", "நான்காயிரம்", "ஐந்தாயிரம்", "ஆறாயிரம்", "ஏழாயிரம்", "எட்டாயிரம்", "ஒன்பதாயிரம்"]
_THOU_CO = ["", "ஆயிரத்து", "இரண்டாயிரத்து", "மூவாயிரத்து", "நான்காயிரத்து", "ஐந்தாயிரத்து", "ஆறாயிரத்து", "ஏழாயிரத்து", "எட்டாயிரத்து", "ஒன்பதாயிரத்து"]


def _two(n):  # 1..99
    if n < 10:
        return DIGIT[n]
    if n < 20:
        return _TEENS[n - 10]
    t, u = divmod(n, 10)
    return _TENS_ST[t] if u == 0 else f"{_TENS_CO[t]} {DIGIT[u]}"


def _three(n):  # 1..999
    h, r = divmod(n, 100)
    if h == 0:
        return _two(r)
    if r == 0:
        return _HUND_ST[h]
    return f"{_HUND_CO[h]} {_two(r)}"


def cardinal(n: int) -> str:
    if n == 0:
        return "பூஜ்ஜியம்"
    if n < 0:
        return "கழித்தல் " + cardinal(-n)
    crore, n = divmod(n, 10_000_000)
    lakh, n = divmod(n, 100_000)
    thou, hund = divmod(n, 1000)
    parts = []
    if crore:
        parts.append(("ஒரு" if crore == 1 else cardinal(crore)) + " கோடி")
    if lakh:
        parts.append(("ஒரு" if lakh == 1 else _three(lakh)) + " லட்சம்")
    if thou:
        if thou < 10:
            parts.append(_THOU_CO[thou] if hund else _THOU_ST[thou])
        else:
            parts.append(_three(thou) + (" ஆயிரத்து" if hund else " ஆயிரம்"))
    if hund:
        parts.append(_three(hund))
    return " ".join(parts)

SYMBOL = {
    "%": " சதவீதம் ", "₹": " ரூபாய் ", "$": " டாலர் ", "&": " மற்றும் ",
    "+": " கூட்டல் ", "=": " சமம் ", "@": " அட்டு ", "°": " டிகிரி ",
}

# a run of latin letters with optional INTERIOR dots: "IIT", "I.I.T", "MLX".
# (a trailing sentence period is left out, so it stays as end-of-sentence punctuation)
_LATIN_RUN = re.compile(r"[A-Za-z](?:\.[A-Za-z])*")
_DIGITS = re.compile(r"\d+")


def _spell_latin(run: str) -> str:
    letters = [c for c in run if c.isalpha()]
    if not letters:
        return run
    # only spell acronym-like runs (all-caps, or originally dotted); leave real words alone
    dotted = "." in run
    all_caps = run.replace(".", "").isupper()
    if not (dotted or all_caps) or len(letters) > 6:
        return run
    return "".join(LETTER.get(c.lower(), "") for c in letters)


def _spell_digits(m: re.Match) -> str:
    s = m.group(0)
    # leading-zero runs (codes, phone numbers) or huge numbers -> read digit by digit
    if (len(s) > 1 and s[0] == "0") or int(s) > 99_99_99_999:
        return " " + " ".join(DIGIT[int(d)] for d in s) + " "
    return " " + cardinal(int(s)) + " "


def normalize(text: str) -> str:
    for sym, word in SYMBOL.items():
        text = text.replace(sym, word)
    text = _LATIN_RUN.sub(lambda m: _spell_latin(m.group(0)), text)
    text = _DIGITS.sub(_spell_digits, text)
    text = re.sub(r"\s+", " ", text).strip()
    return text


if __name__ == "__main__":
    for t in ["IIT", "I.I.T.", "MLX model", "காலை உணவு", "100% சரி", "2026 ஆம் ஆண்டு", "₹50"]:
        print(f"{t!r:30} -> {normalize(t)!r}")

File: rust/src/normalize.rs
//! Text normalization (verbalization) — mirror of `tamiltts/mlx/normalize.py`. Keep in sync.
//!
//! Run before char tokenization: acronyms/Latin-letter runs -> Tamil letter-names (periods dropped),
//! symbols -> Tamil words, digit runs -> Tamil cardinal number-words (digit-by-digit for leading-zero
//! codes or huge numbers).

use regex::{Captures, Regex};

const DIGIT: [&str; 10] = ["பூஜ்ஜியம்", "ஒன்று", "இரண்டு", "மூன்று", "நான்கு", "ஐந்து", "ஆறு", "ஏழு", "எட்டு", "ஒன்பது"];
const TEENS: [&str; 10] = ["பத்து", "பதினொன்று", "பன்னிரண்டு", "பதிமூன்று", "பதினான்கு", "பதினைந்து", "பதினாறு", "பதினேழு", "பதினெட்டு", "பத்தொன்பது"];
const TENS_ST: [&str; 10] = ["", "பத்து", "இருபது", "முப்பது", "நாற்பது", "ஐம்பது", "அறுபது", "எழுபது", "எண்பது", "தொண்ணூறு"];
const TENS_CO: [&str; 10] = ["", "", "இருபத்தி", "முப்பத்தி", "நாற்பத்தி", "ஐம்பத்தி", "அறுபத்தி", "எழுபத்தி", "எண்பத்தி", "தொண்ணூற்றி"];
const HUND_ST: [&str; 10] = ["", "நூறு", "இருநூறு", "முந்நூறு", "நானூறு", "ஐந்நூறு", "அறுநூறு", "எழுநூறு", "எண்ணூறு", "தொள்ளாயிரம்"];
const HUND_CO: [&str; 10] = ["", "நூற்றி", "இருநூற்று", "முந்நூற்று", "நானூற்று", "ஐந்நூற்று", "அறுநூற்று", "எழுநூற்று", "எண்ணூற்று", "தொள்ளாயிரத்து"];
const THOU_ST: [&str; 10] = ["", "ஆயிரம்", "இரண்டாயிரம்", "மூவாயிரம்", "நான்காயிரம்", "ஐந்தாயிரம்", "ஆறாயிரம்", "ஏழாயிரம்", "எட்டாயிரம்", "ஒன்பதாயிரம்"];
const THOU_CO: [&str; 10] = ["", "ஆயிரத்து", "இரண்டாயிரத்து", "மூவாயிரத்து", "நான்காயிரத்து", "ஐந்தாயிரத்து", "ஆறாயிரத்து", "ஏழாயிரத்து", "எட்டாயிரத்து", "ஒன்பதாயிரத்து"];

const SYMBOLS: [(&str, &str); 8] = [
    ("%", " சதவீதம் "), ("₹", " ரூபாய் "), ("$", " டாலர் "), ("&", " மற்றும் "),
    ("+", " கூட்டல் "), ("=", " சமம் "), ("@", " அட்டு "), ("°", " டிகிரி "),
];

fn letter_name(c: char) -> &'static str {
    match c.to_ascii_lowercase() {
        'a' => "ஏ", 'b' => "பி", 'c' => "சி", 'd' => "டி", 'e' => "ஈ", 'f' => "எஃப்", 'g' => "ஜி",
        'h' => "ஹெச்", 'i' => "ஐ", 'j' => "ஜே", 'k' => "கே", 'l' => "எல்", 'm' => "எம்", 'n' => "என்",
        'o' => "ஓ", 'p' => "பி", 'q' => "கியூ", 'r' => "ஆர்", 's' => "எஸ்", 't' => "டி", 'u' => "யூ",
        'v' => "வி", 'w' => "டபிள்யூ", 'x' => "எக்ஸ்", 'y' => "வை", 'z' => "செட்", _ => "",
    }
}

fn two(n: u64) -> String {
    if n < 10 { return DIGIT[n as usize].into(); }
    if n < 20 { return TEENS[(n - 10) as usize].into(); }
    let (t, u) = (n / 10, n % 10);
    if u == 0 { TENS_ST[t as usize].into() } else { format!("{} {}", TENS_CO[t as usize], DIGIT[u as usize]) }
}

fn three(n: u64) -> String {
    let (h, r) = (n / 100, n % 100);
    if h == 0 { return two(r); }
    if r == 0 { return HUND_ST[h as usize].into(); }
    format!("{} {}", HUND_CO[h as usize], two(r))
}

fn cardinal(n: u64) -> String {
    if n == 0 { return "பூஜ்ஜியம்".into(); }
    let crore = n / 10_000_000;
    let n = n % 10_000_000;
    let lakh = n / 100_000;
    let n = n % 100_000;
    let (thou, hund) = (n / 1000, n % 1000);
    let mut parts: Vec<String> = Vec::new();
    if crore > 0 { parts.push(format!("{} கோடி", if crore == 1 { "ஒரு".into() } else { cardinal(crore) })); }
    if lakh > 0 { parts.push(format!("{} லட்சம்", if lakh == 1 { "ஒரு".into() } else { three(lakh) })); }
    if thou > 0 {
        if thou < 10 {
            parts.push((if hund > 0 { THOU_CO[thou as usize] } else { THOU_ST[thou as usize] }).into());
        } else {
            parts.push(format!("{}{}", three(thou), if hund > 0 { " ஆயிரத்து" } else { " ஆயிரம்" }));
        }
    }
    if hund > 0 { parts.push(three(hund)); }
    parts.join(" ")
}

fn spell_latin(run: &str) -> String {
    let letters: Vec<char> = run.chars().filter(|c| c.is_ascii_alphabetic()).collect();
    if letters.is_empty() { return run.to_string(); }
    let dotted = run.contains('.');
    let all_caps = letters.iter().all(|c| c.is_ascii_uppercase());
    if !(dotted || all_caps) || letters.len() > 6 { return run.to_string(); }
    letters.iter().map(|&c| letter_name(c)).collect()
}

fn spell_digits(s: &str) -> String {
    let n: u64 = s.parse().unwrap_or(u64::MAX);
    if (s.len() > 1 && s.starts_with('0')) || n > 99_99_99_999 {
        let d: Vec<&str> = s.chars().map(|c| DIGIT[c as usize - '0' as usize]).collect();
        return format!(" {} ", d.join(" "));
    }
    format!(" {} ", cardinal(n))
}

pub fn normalize(text: &str) -> String {
    let mut s = text.to_string();
    for (sym, word) in SYMBOLS { s = s.replace(sym, word); }
    let latin = Regex::new(r"[A-Za-z](?:\.[A-Za-z])*").unwrap();
    s = latin.replace_all(&s, |c: &Captures| spell_latin(&c[0])).into_owned();
    let digits = Regex::new(r"\d+").unwrap();
    s = digits.replace_all(&s, |c: &Captures| spell_digits(&c[0])).into_owned();
    s.split_whitespace().collect::<Vec<_>>().join(" ")
}
